### PR TITLE
Oracle PLSQL lexer (suitable for Oracle SQL as well)

### PR DESCRIPTION
**Pull request by lee-lindley**
The keywords, non-reserved keywords, types and built-in functions are from the Oracle 19c manuals.
The quoted literal parsing is much better than the existing SQL lexer, especially regarding the "q'" operator. I classify the quote delimiters as operators which is a departure from most others, but they are not part of the literal, so making them the same color as the quoted string is just wrong.

I borrowed liberally from the existing SQL, Perl and Ruby lexers and folowed the Lexer Development guide.
The demo, sample and spec files are similar to others in those directories.

My blog (http://lee-lindley.github.io) is built using this lexer.

### CODE REVIEW

File: lib/rouge/lexers/plsql.rb [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+
+      state :double_string do
+        rule %r/\\./, Str::Escape
+        rule %r/""/, Str::Escape
+        rule %r/"/, Operator, :pop!
+        rule %r/[^\\"]+/m, Name::Variable
+      end
+
+      state :dotnames do
+            # if we are followed by a dot and another name, we are an ordinairy name
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:576]: **nitpick:** Typo?

```suggestion
            # if we are followed by a dot and another name, we are an ordinary name
```

File: lib/rouge/lexers/plsql.rb [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+        OCIROWID OCISTRING OCITYPE OF ON OPTION OR ORACLE ORADATA ORDER ORLANY ORLVARY
+        OUT OVERRIDING PARALLEL_ENABLE PARAMETER PASCAL PCTFREE PIPE PIPELINED POLYMORPHIC
+        PRAGMA PRIOR PUBLIC RAISE RECORD RELIES_ON REM RENAME RESOURCE RESULT REVOKE ROWID 
+        SB1 SB2 SELECT SEPARATE SET SHARE SHORT SIZE SIZE_T SPARSE SQLCODE SQLDATA
+        SQLNAME SQLSTATE STANDARD START STORED STRUCT STYLE SYNONYM TABLE TDO THEN
+        TRANSACTIONAL TRIGGER UB1 UB4 UNION UNIQUE UNSIGNED UNTRUSTED UPDATE VALIST
+        VALUES VARIABLE VIEW VOID WHERE WHILE WITH
+        ))
+      end
+      def self.keywords_nresvd
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:32]: **nitpick:** Could we add a line separator between these two functions for better readability?

Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:32]: **question:** Can we rename this method to `keywords_reserved` for consistency with the other lexers? Or does this name refer to something totally different? 🤔

File: lib/rouge/lexers/plsql.rb [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+        WITH_PLSQL WORK WRAPPED WRAPPER WRITE XDB_FASTPATH_INSERT XID XML XML2OBJECT
+        XMLATTRIBUTES XMLCAST XMLCDATA XMLCOLATTVAL XMLCOMMENT XMLCONCAT XMLDIFF XMLELEMENT
+        XMLEXISTS XMLEXISTS2 XMLFOREST XMLINDEX_REWRITE XMLINDEX_REWRITE_IN_SELECT
+        XMLINDEX_SEL_IDX_TBL XMLISNODE XMLISVALID XMLNAMESPACES XMLPARSE XMLPATCH
+        XMLPI XMLQUERY XMLQUERYVAL XMLROOT XMLSCHEMA XMLSERIALIZE XMLTABLE XMLTOJSON
+        XMLTOKENSET XMLTRANSFORM XMLTRANSFORMBLOB XMLTSET_DML_ENABLE XML_DIAG XML_DML_RWT_STMT
+        XPATHTABLE XS XS_SYS_CONTEXT X_DYN_PRUNE YEARS YES ZONEMAP
+        ))
+      end
+      def self.keywords_func
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:390]: **nitpick:** Same as above. Could we add a blank line between function blocks?

File: lib/rouge/lexers/plsql.rb [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+        TRANSLATE TREAT TRIM TRUNC TZ_OFFSET UID UNISTR UPPER USER USERENV VALIDATE_CONVERSION
+        VALUE VARIANCE VAR_POP VAR_SAMP VSIZE WIDTH_BUCKET XMLAGG XMLCAST XMLCDATA
+        XMLCOLATTVAL XMLCOMMENT XMLCONCAT XMLDIFF XMLELEMENT XMLEXISTS XMLFOREST XMLISVALID
+        XMLPARSE XMLPATCH XMLPI XMLQUERY XMLROOT XMLSEQUENCE XMLSERIALIZE XMLTABLE
+        XMLTRANSFORM
+        ))
+      end
+
+      def self.keywords_type
+        # sources:
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:433]: **suggestion:** Is this comment needed?

File: lib/rouge/lexers/plsql.rb [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+        # eat whitespace including newlines
+        rule %r/\s+/m, Text
+
+        #
+        # Comments
+        #
+        rule %r/--.*/, Comment::Single
+        rule %r(/\*), Comment::Multiline, :multiline_comments
+        # Q' operator quoted string literal
+        rule %r/q'(.)/i  do |m|
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:462]:
```suggestion
        rule %r/q'(.)/i do |m|
```

File: lib/rouge/lexers/plsql.rb [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+            #open = Regexp.escape(m[1])
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:463]: **question:** Redundant comment?

File: lib/rouge/lexers/plsql.rb [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+        end
+        rule %r/'/, Operator, :single_string
+        # A double-quoted string refers to a database object in our default SQL
+        rule %r/"/, Operator, :double_string
+        rule %r/(\$(?:IF|THEN|ELSE|ELSIF|ERROR|END|(?:\$\$?\w[\w\d]*)))(\s+)/im do
+            groups Comment::Preproc, Text
+        end
+        ### we do not use backticks in Oracle. I do not know the rules for other sql engines
+        ###rule %r/`/, Name::Variable, :backtick
+
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:480]: **suggestion:** I think we could drop these comments. What do you think?

File: lib/rouge/lexers/plsql.rb [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+        end
+
+        #
+        # General keyword classification with sepcial attention to names
+        # in a chained "dot" notation.
+        #
+        rule %r/(\w[\w\d\$]*)(\.(?=\w))?/ do |m|
+          if self.class.keywords_type.include? m[1].upcase
+            tok = Keyword::Type 
+            #Name::Builtin
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:530]: **question:** Redundant comment?

File: lib/rouge/lexers/plsql.rb [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+          elsif self.class.keywords_func.include? m[1].upcase
+            tok = Name::Function
+          elsif self.class.keywords.include? m[1].upcase
+            tok = Keyword::Reserved
+          elsif self.class.keywords_nresvd.include? m[1].upcase
+            tok = Keyword
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:536]: **suggestion:** Could we swap the token types?

```suggestion
          elsif self.class.keywords.include? m[1].upcase
            tok = Keyword
          elsif self.class.keywords_nresvd.include? m[1].upcase
            tok = Keyword::Reserved
```

Or I could have misread the intention here.

File: lib/rouge/lexers/plsql.rb [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        rule %r([*]\/), Comment::Multiline, :pop!
+      end
+
+
+      #state :backtick do
+      #  rule %r/\\./, Str::Escape
+      #  rule %r/``/, Str::Escape
+      #  rule %r/`/, Name::Variable, :pop!
+      #  rule %r/[^\\`]+/, Name::Variable
+      #end
```
Comment (tancnle) [inline lib/rouge/lexers/plsql.rb:559]: **suggestion:** If backtick is not supported in PL/SQL, can we remove this block altogether instead of commenting?

**Review by tancnle — COMMENTED** [review]
Thank you for your contribution and hard work @lee-lindley ❤️ The PR looks great to me overall 👍🏼 I have added some suggestions/questions for your consideration.

- Could we ensure 2 space indentation on the `lib/rouge/lexers/psql.rb` file? There seem to be some mixed 2 and 4 space indentations on the file.
- Could we add an entry on the `docs/Languages.md` file for this new lexer?

Please let me know what you think 🙏🏼 .

File: lib/rouge/lexers/plsql.rb (hunk 10 above)
Comment (lee-lindley) [reply lib/rouge/lexers/plsql.rb:559]: Thanks for the detailed review Tan!
I'll address all of this as time permits. It shouldn't take long (though not today).
Unfortunately, if I dive back in I'll be tempted to tweak some more. I'll try to restrain myself.

**Comment by lee-lindley** [comment]
Tan,
I answered each one as I was editing and pressed the "resolved" button. Because I do not see them here, I'm thinking github dumped my comments when I did that instead of saving it along with the resolution. Lesson learned.

In any case I believe I addressed all suggestions appropriately, and only did small amount of functionality tweaking. I reviewed my test files visually and everything still looks good.

**Comment by tancnle** [comment]
Thanks a lot for your prompt action @lee-lindley 🙇🏼 Looks great to me 🚀

**Pull request merged**